psi4/src/psi4/libmints/thc_eri.cc

```diff
@@ -0,0 +1,354 @@
+    nthreads = Process::environment.get_n_threads();
+#endif
+
+    IntegralFactory factory(primary_);
+    std::vector<std::shared_ptr<TwoBodyAOInt>> eri_computers(nthreads);
+
+    eri_computers[0] = std::shared_ptr<TwoBodyAOInt>(factory.eri());
+#pragma omp parallel for
+    for (int thread = 1; thread < nthreads; ++thread) {
+        eri_computers[thread] = std::shared_ptr<TwoBodyAOInt>(eri_computers[0]->clone());
+    }
+
+    size_t nshellpair = eri_computers[0]->shell_pairs().size();
+
+    SharedMatrix E_PQ = std::make_shared<Matrix>(rank, rank);
+
+#pragma omp parallel for
+    for (size_t MN = 0; MN < nshellpair; ++MN) {
+
+        int thread = 0;
+#ifdef _OPENMP
+        thread = omp_get_thread_num();
+#endif
+
+        auto bra = eri_computers[thread]->shell_pairs()[MN];
+        size_t M = bra.first;
+        size_t N = bra.second;
+
+        int nm = primary_->shell(M).nfunction();
+        int mstart = primary_->shell(M).function_index();
+        int nn = primary_->shell(N).nfunction();
+        int nstart = primary_->shell(N).function_index();
+
+        SharedMatrix E_temp = std::make_shared<Matrix>(rank, nm * nn);
+
+        double prefactor1 = (M == N) ? 1.0 : 2.0;
+
+        for (size_t RS = 0; RS < nshellpair; ++RS) {
+
+            auto ket = eri_computers[thread]->shell_pairs()[RS];
+            size_t R = ket.first;
+            size_t S = ket.second;
+
+            int nr = primary_->shell(R).nfunction();
+            int rstart = primary_->shell(R).function_index();
+            int ns = primary_->shell(S).nfunction();
+            int sstart = primary_->shell(S).function_index();
+
+            // TODO: Implement screening to make this more efficient
+            eri_computers[thread]->compute_shell(M, N, R, S);
+            const auto &buffer = eri_computers[thread]->buffers()[0];
+
+            double prefactor2 = (R == S) ? 1.0 : 2.0;
+
+            for (size_t p = 0; p < rank; ++p) {
+                for (size_t dm = 0, index = 0; dm < nm; ++dm) {
+                    for (size_t dn = 0; dn < nn; ++dn) {
+                        for (size_t r = rstart; r < rstart + nr; ++r) {
+                            for (size_t s = sstart; s < sstart + ns; ++s, ++index) {
+                                (*E_temp)(p, dm * nn + dn) += prefactor2 * buffer[index] * (*x1_)(p, r) * (*x1_)(p, s);
+                            } // end s
+                        } // end r
+                    } // end dn
+                } // end dm
+            } // end p
+        } // end RS
+
+        for (size_t p = 0; p < rank; ++p) {
+            for (size_t q = 0; q < rank; ++q) {
+                for (size_t m = mstart; m < mstart + nm; ++m) {
+                    size_t dm = m - mstart;
+                    for (size_t n = nstart; n < nstart + nn; ++n) {
+                        size_t dn = n - nstart;
+    #pragma omp atomic
+                        (*E_PQ)(p, q) += prefactor1 * (*E_temp)(q, dm * nn + dn) * (*x1_)(p, m) * (*x1_)(p, n);
```

Comment:
Would it be a good idea to accumulate pq contribution first before adding it to (*E_PQ)(p, q). That way you only have pq atomic operations instead of one for every pqmn?

Reply:
Yes, good point